pkg/apis/v1beta1/nodeclass.go

```diff
@@ -0,0 +1,278 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/hashstructure/v2"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NodeClassSpec is the top level specification for the AWS Karpenter Provider.
+// This will contain configuration necessary to launch instances in AWS.
+type NodeClassSpec struct {
+	// AMIFamily is the AMI family that instances use.
+	// +optional
+	AMIFamily *string `json:"amiFamily,omitempty"`
+	// UserData to be applied to the provisioned nodes.
+	// It must be in the appropriate format based on the AMIFamily in use. Karpenter will merge certain fields into
+	// this UserData to ensure nodes are being provisioned with the correct configuration.
+	// +optional
+	UserData *string `json:"userData,omitempty"`
+	// Role is the AWS identity that nodes use.
+	// +optional
+	Role *string `json:"role,omitempty"`
+	// SubnetSelectorTerms is a list of or subnet selector terms. The terms are ORed.
+	// +optional
+	SubnetSelectorTerms []SubnetSelectorTerm `json:"subnetSelectorTerms" hash:"ignore"`
+	// SecurityGroupSelectorTerms is a list of or security group selector terms. The terms are ORed.
+	// +optional
+	SecurityGroupSelectorTerms []SecurityGroupSelectorTerm `json:"securityGroupSelectorTerms" hash:"ignore"`
+	// AMISelectorTerms is a list of or ami selector terms. The terms are ORed.
+	// +optional
+	AMISelectorTerms []AMISelectorTerm `json:"amiSelectorTerms,omitempty" hash:"ignore"`
+	// Context is a Reserved field in EC2 APIs
+	// https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateFleet.html
+	// +optional
+	Context *string `json:"context,omitempty"`
+	// Tags to be applied on ec2 resources like instances and launch templates.
+	// +optional
+	Tags map[string]string `json:"tags,omitempty"`
```

Comment:
How did we do the ordering of these? For folks reading, you could imagine hoisting this one higher up. Or we could alphabetize.

Reply:
This is a good point. I think we want to do it by how common the fields are along with things that should be naturally grouped together. I don't think that alphabetic makes a lot of sense on its own.